lib/puppet/provider/nameservice.rb

```diff
@@ -288,5 +288,25 @@ def set(param, value)
       raise Puppet::Error, "Could not set #{param} on #{@resource.class.name}[#{@resource.name}]: #{detail}", detail.backtrace
     end
   end
+
+  # From overriding Puppet::Property#insync?
+  # Ruby Etc::getpwnam < 2.1.0 always returns a struct with binary encoded
+  # string values, and >= 2.1.0 will return binary encoded strings for values
+  # incompatible with current locale charset. Convert "should" value to binary
+  # if "current" value is, to avoid unnecessary property syncs and concatenation
+  # of strings with incompatible encodings. (PUP-6777)
+  #
+  # * "should" parameter will always be a string with one possible (strictly
+  # speaking) exception, which is the boolean value true. In current semantics,
+  # this would convert to a string. For strict backwards-compatibility, convert
+  # it to a string if this is the case.
+  #
+  # return basic string comparison after encoding (same as
+  # Puppet::Property#property_matches)
+  def comments_insync?(current, should)
+    should = should.to_s if should == true
+    should.force_encoding(Encoding::ASCII_8BIT) if current.encoding == Encoding::ASCII_8BIT
```

Comment:
So your question is an interesting one @MosesMendoza.  When encodings mismatch, I think we have 2 potential ways we can fail:

* `current == should` explodes on the mismatched encodings
* On the attempt to match the encodings, a conversion error is raised at the `force_encoding` (let's say this is some weird case where Ruby doesn't give us a binary string - which you seem to indicate isn't possible)

Which propagates a more useful error message?